flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/ElasticsearchSinkBase.java

```diff
@@ -0,0 +1,234 @@
+
+	/** Set from inside the @link BulkProcessor} listener if a {@link Throwable} was thrown during processing. */
+	private final AtomicReference<Throwable> failureThrowable = new AtomicReference<>();
+
+	public ElasticsearchSinkBase(ElasticsearchClientFactory clientFactory,
+								Map<String, String> userConfig,
+								ElasticsearchSinkFunction<T> elasticsearchSinkFunction) {
+		this.clientFactory = checkNotNull(clientFactory);
+		this.elasticsearchSinkFunction = checkNotNull(elasticsearchSinkFunction);
+
+		// we eagerly check if the user-provided sink function is serializable;
+		// otherwise, if it isn't serializable, users will merely get a non-informative error message
+		// "ElasticsearchSinkBase is not serializable"
+		try {
+			InstantiationUtil.serializeObject(elasticsearchSinkFunction);
+		} catch (Exception e) {
+			throw new IllegalArgumentException(
+				"The implementation of the provided ElasticsearchSinkFunction is not serializable. " +
+				"The object probably contains or references non serializable fields.");
+		}
+
+		checkNotNull(userConfig);
+
+		// extract and remove bulk processor related configuration from the user-provided config,
+		// so that the resulting user config only contains configuration related to the Elasticsearch client.
+		ParameterTool params = ParameterTool.fromMap(userConfig);
+
+		if (params.has(CONFIG_KEY_BULK_FLUSH_MAX_ACTIONS)) {
+			bulkProcessorFlushMaxActions = params.getInt(CONFIG_KEY_BULK_FLUSH_MAX_ACTIONS);
+			userConfig.remove(CONFIG_KEY_BULK_FLUSH_MAX_ACTIONS);
+		} else {
+			bulkProcessorFlushMaxActions = null;
+		}
+
+		if (params.has(CONFIG_KEY_BULK_FLUSH_MAX_SIZE_MB)) {
+			bulkProcessorFlushMaxSizeMb = params.getInt(CONFIG_KEY_BULK_FLUSH_MAX_SIZE_MB);
+			userConfig.remove(CONFIG_KEY_BULK_FLUSH_MAX_SIZE_MB);
+		} else {
+			bulkProcessorFlushMaxSizeMb = null;
+		}
+
+		if (params.has(CONFIG_KEY_BULK_FLUSH_INTERVAL_MS)) {
+			bulkProcessorFlushIntervalMillis = params.getInt(CONFIG_KEY_BULK_FLUSH_INTERVAL_MS);
+			userConfig.remove(CONFIG_KEY_BULK_FLUSH_INTERVAL_MS);
+		} else {
+			bulkProcessorFlushIntervalMillis = null;
+		}
+
+		this.userConfig = userConfig;
+	}
+
+	@Override
+	public void open(Configuration parameters) throws Exception {
+		client = clientFactory.create(userConfig);
+
+		BulkProcessor.Builder bulkProcessorBuilder = BulkProcessor.builder(
+			client,
+			new BulkProcessor.Listener() {
+				@Override
+				public void beforeBulk(long executionId, BulkRequest request) { }
+
+				@Override
+				public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
+					if (response.hasFailures()) {
+						for (BulkItemResponse itemResp : response.getItems()) {
+							if (itemResp.isFailed()) {
+								LOG.error("Failed to index document in Elasticsearch: " + itemResp.getFailureMessage());
+								failureThrowable.compareAndSet(null, new RuntimeException(itemResp.getFailureMessage()));
```

Comment:
Could be replaced by ` failureThrowable.compareAndSet(null, itemResp.getFailure().getCause());`